### PR TITLE
Feature/issue 574 invalid delimited import

### PR DESCRIPTION
**Pull request by serpens24**
### Description of the Change

Handle potential errors that could occur when importing delimited files, these are grouped into 3 categories:

1. Invalid files are selected to add to source files.
2. Files that are already listed in source pane become invalid or are deleted and are selected by user in source pane.
3. Files that are already listed in source pane become invalid or are deleted and user presses import button.

The basic approahc is to try and elimiate invalid files as early as possible, but still be aware that files can change subsequent to being selected.

Rather than letting unhandled exceptions bubble up to the plugin code, catch these invalid files and use neat error dialogs that come to the front of screen, rather than being hidden under the Import Delimited File dialog.
Ie:

WarningImport
SuccessfulImport
SelectMissingOrInvalidFileAlreadyInSourceList

### Alternate Designs

Leaving code as is and allowign errors to be thrown.

### Why Should This Be In Core?

Better user experience for users importing delimited files.

### Benefits

As above

### Possible Drawbacks

None known

### Verification Process

Testing of the following:
1. Invalid files are selected to add to source files.
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select one valid and one invalid file.
* confirm a warning dialog is displayed, and that after pressong OK, only the valid file is added to the list in the source pane.

2. Files that are already listed in source pane become invalid or are deleted and are selected by user in source pane.|
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select two valid files.
* Confirm no errors are detected and both appear in the source file list.
* On disk delete or invalidate the contents of the file NOT being previewed in the source pane. 
* Select the now invalid (or missing) file in the source pane and confirm an warning dialog is displayed and that the file is removed from the source pane

3a. Files that are already listed in source pane become invalid or are deleted and user presses import button.
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select two valid files.
* Confirm no errors are detected and both appear in the source file list.
* Map one or more fileds in the action pane.
* Invalidate or delete one of the files on disk and press the Import button.
* Confirm a warning dialog is shown and no exception dialogs thrown. 

3b. Files that are already listed in source pane become invalid or are deleted and user presses import button.
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select two valid files.
* Confirm no errors are detected and both appear in the source file list.
* Map one or more fileds in the action pane.
* Confirm an info dialog is shown documenting number of rows imported. (Note rows could be duplicated, so number of data rows does not always equate to number of nodes/edges).
### Applicable Issues

https://github.com/constellation-app/constellation/issues/574
https://github.com/constellation-app/constellation/issues/606

### CODE REVIEW

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 1]
```diff
@@ -449,9 +470,19 @@ private void updateSampleData() {
                 currentColumns = new String[columns.length + 1];
                 System.arraycopy(columns, 0, currentColumns, 1, columns.length);
                 currentColumns[0] = "Row";
-            } catch (IOException ex) {
-                final NotifyDescriptor nd = new NotifyDescriptor.Message(ex.getMessage(), NotifyDescriptor.WARNING_MESSAGE);
-                DialogDisplayer.getDefault().notify(nd);
+            } catch (FileNotFoundException ex) {
+                String warningMsg = "The following file could not be found and has been excluded from the import set:\n  " + sampleFile.getPath();
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:474]: Add Final String warningMsg

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 2]
```diff
@@ -449,9 +470,19 @@ private void updateSampleData() {
...
+                LOGGER.log(Level.INFO, warningMsg);
+                displayAlert("Invalid file selected", warningMsg, Alert.AlertType.WARNING);
+                files.remove(sampleFile);
+                stage.getSourcePane().removeFile(sampleFile);
+            }
+            catch (IOException ex) {
+                String warningMsg = "The following file could not be parsed and has been excluded from the import set:\n  " + sampleFile.getPath();
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:481]: Add a final

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java [hunk 3]
```diff
@@ -112,6 +115,58 @@ public PluginParameters createParameters() {
...
+    /**
+     * Build up an import summary dialog detailing successful and unsuccessful file imports.
+     * 
+     * @param title Title to add to status dialog
+     * @param importedRows Number of rows successfully imported (from valid files)
+     * @param validFilenames List of filenames that were imported from
+     * @param invalidFilenames List of files that couldn't be opened/parsed. We try to limit this possibility by pre-screening
+     *                         files during the initial file selection.
+     */
+    private void displaySummaryAlert(int importedRows, List<String> validFilenames, List<String> invalidFilenames) {
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java:128]: Make parameters final

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java [hunk 4]
```diff
@@ -124,33 +179,50 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
         final PluginParameters parserParameters = (PluginParameters) parameters.getParameters().get(PARSER_PARAMETER_IDS_PARAMETER_ID).getObjectValue();
         final List<Integer> newVertices = new ArrayList<>();
         boolean positionalAtrributesExist = false;
+        List<String> validFiles = new ArrayList<>();
+        List<String> invalidFiles = new ArrayList<>();
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java:183]: Make both lists final

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java [hunk 5]
```diff
@@ -124,33 +179,50 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
             interaction.setProgress(0, 0, "Reading File: " + file.getName(), true);
             List<String[]> data = null;
+            
             try {
                 data = parser.parse(new InputSource(file), parserParameters);
+                importRows = importRows + data.size() - 1;
+                validFiles.add(file.getPath());
+                LOGGER.log(Level.INFO, "Imported {0} rows of data from file {1}. {2} total rows imported", new Object[]{(data.size() - 1), file.getPath(), importRows});
+            } catch (FileNotFoundException ex) {
+                String errorMsg = file.getPath() + " could not be found. Ignoring file during import.";
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java:196]: final

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java [hunk 6]
```diff
@@ -124,33 +179,50 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
+                LOGGER.log(Level.INFO, errorMsg);
+                invalidFiles.add(file.getPath());
             } catch (IOException ex) {
-                throw new PluginException(this, PluginNotificationLevel.ERROR, "Error reading file: " + file.getName(), ex);
+                String errorMsg = file.getPath() + " could not be parsed. Removing file during import.";
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedPlugin.java:200]: final

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/SourcePane.java [hunk 7]
```diff
@@ -155,11 +161,24 @@ public SourcePane(final ImportController importController) {
                     SourcePane.this.importFileParserComboBox.setDisable(true);
                 }
                 ObservableList<File> files = FXCollections.observableArrayList(fileListView.getItems());
+                String warningMsg = "The following files could not be parsed and have been excluded from import set:";
+                boolean foundInvalidFile = false;
                 for (File file : newFiles) {
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/SourcePane.java:166]: May as well add a final here
```
for (final File file : newFiles) {
```

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/SourcePane.java [hunk 8]
```diff
@@ -155,11 +161,24 @@ public SourcePane(final ImportController importController) {
                     SourcePane.this.importFileParserComboBox.setDisable(true);
                 }
                 ObservableList<File> files = FXCollections.observableArrayList(fileListView.getItems());
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/SourcePane.java:163]: I know this isn't your change but maybe adding a final here?

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/SourcePane.java [hunk 9]
```diff
@@ -252,6 +268,16 @@ public void setParameters(final PluginParameters parameters) {
             parametersPane.getChildren().add(pluginParametersPane);
         }
     }
+    
+    /** Allow a file to be removed from fileListView. This would be triggered by code in InputController if the file
+     * was found to be missing or invalid - these checks are triggered when a new file is selected in the fileListView.
+     * @param file The file to remove.
+     */
+    public void removeFile(File file) {
+        ObservableList<File> files = fileListView.getItems();
```
Comment (aldebaran30701) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/SourcePane.java:277]: final

**Comment by arcturus2** [comment]
I noticed that with the json files the dialog was not showing the file name.

This is after adding an invalid json file:
image

And a valid one:
image

I'm doing this from Linux so not sure if that is the reason. I'm happy to merge this in to include this enhancement before the code freeze.

**Comment by formalhaut69** [comment]
> I noticed that with the json files the dialog was not showing the file name.
> 
> I'm doing this from Linux so not sure if that is the reason. I'm happy to merge this in to include this enhancement before the code freeze.

@arcturus2, @serpens24 the file names appear on my machine. Should I make a ticket that addresses the file names not showing on Linux?

image

image

**Pull request merged**